r-package/grf/R/boosted_regression_forest.R

```diff
@@ -0,0 +1,248 @@
+    #save tuned parameters for use on future boosting iterations
+    tunable.params <- forest.Y$tunable.params
+    Y.hat <- forest.Y.p$prediction
+    error.debiased <- forest.Y.p$debiased.error
+    boosted.forest$forests[[1]] <- forest.Y
+
+    if(!is.null(num.steps)) {
+        for (step in 2:num.steps) {
+            E.hat <- Y - Y.hat
+            forest.E <- regression_forest(X,E.hat, sample.fraction = tunable.params["sample.fraction"],
+                                        mtry = tunable.params["mtry"], tune.parameters = FALSE,
+                                          num.trees = num.trees,
+                                           num.threads = num.threads,
+                                           min.node.size = tunable.params["min.node.size"],
+                                           honesty = honesty,
+                                          honesty.fraction = honesty.fraction,
+                                           seed = seed, ci.group.size = ci.group.size,
+                                          alpha = tunable.params["alpha"],
+                                          imbalance.penalty = tunable.params["imbalance.penalty"],
+                                        clusters = clusters, samples_per_cluster = samples_per_cluster);
+            current.pred <- predict(forest.E)
+            Y.hat <- Y.hat + current.pred$predictions
+            error.debiased <- current.pred$debiased.error
+            boosted.forest[["forests"]][[step]] <- forest.E
+        }
+    }
+    else {
+        step <- 2
+        tolerance <- 0.99
+        still_boosting <- TRUE
+        while(still_boosting & step <= max.steps) {
+            E.hat <- Y - Y.hat
+            forest.small <- regression_forest(X,E.hat,
+                                    sample.fraction = as.numeric(tunable.params["sample.fraction"]),
+                                        mtry = as.numeric(tunable.params["mtry"]), tune.parameters = FALSE,
+                                          num.trees = num.trees.tune,
+                                           num.threads = num.threads,
+                                           min.node.size = as.numeric(tunable.params["min.node.size"]),
+                                           honesty = honesty,
+                                          honesty.fraction = honesty.fraction,
+                                           seed = seed, ci.group.size = ci.group.size,
+                                          alpha = as.numeric(tunable.params["alpha"]),
+                                          imbalance.penalty = as.numeric(tunable.params["imbalance.penalty"]),
+                                        clusters = clusters, samples_per_cluster = samples_per_cluster);
+            step.error.approx <- predict(forest.small)$debiased.error
+            if((mean(step.error.approx,na.rm=TRUE) < tolerance*mean(error.debiased,na.rm=TRUE))) {
+                forest.E <- regression_forest(X,E.hat,
```

Comment:
This is exactly the same thing as lines 116-130. It'd be good to consolidate these two code chunks into a single flow. Something like:
- first compute residuals
- check if we should still do a step of boosting (trivial if num.steps specified, more involved if we're tuning)
- perform a step of boosting if appropriate